### PR TITLE
Remove most of the PTQ receipe

### PR DESCRIPTION
**Pull request by jerryzh168**
Summary:
We want to decouple PTQ and torchtune, to avoid headaches of torchao API changes breaking torchtune. Also curernt PTQ APIs in torchtune is not the most up to date ones, by removing them we could deprecate these API in torchao as well.

8da4w ptq is kept because it's used in QAT, but it can be removed after QAT is refactored as well.

User should use torchao directly after torchtune finetuning: https://github.com/pytorch/ao/blob/main/torchao/quantization/README.md to quantize their model
and do eval/benchmark in torchao as well: https://github.com/pytorch/ao/tree/main/torchao/_models/llama

Test Plan:
CI

Reviewers:

Subscribers:

Tasks:

Tags:

#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [x] other (please add here)

Deprecate old APIs

#### Changelog
What are the changes made in this PR?

#### Test plan
Please make sure to do each of the following if applicable to your PR. (If you're not sure about any one of these just ask and we will happily help.)

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add unit tests for any new functionality
- [ ] update docstrings for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
	- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

### CODE REVIEW

File: recipes/quantization.md [hunk 1]
```diff
@@ -1,14 +1,11 @@
 # Quantization and Sparsity
 
-torchtune integrates with [torchao](https://github.com/pytorch/ao/) for architecture optimization techniques including quantization and sparsity. Currently only some quantization techniques are integrated, see the docstrings in the [quantization recipe](quantize.py) and the [QAT recipe](qat_distributed.py) for more details.
+torchtune integrates with [torchao](https://github.com/pytorch/ao/) for QAT and QLoRA. Currently only some quantization techniques are integrated, see the docstrings in the [quantization recipe](quantize.py) and the [QAT recipe](qat_distributed.py) for more details.
 
-#### Quantize
-To quantize a model (default is int4 weight only quantization):
-```
-tune run quantize --config quantization
```
Comment (andrewor14) [inline recipes/quantization.md:8]: note: this is still referenced in the QAT tutorial, so maybe we need to keep this in README but move it to under the QAT section?

Comment (andrewor14) [inline recipes/quantization.md:8]: note: this is still referenced in the [QAT tutorial](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html), so maybe we need to keep this in README but move it to under the QAT section?

File: recipes/quantization.md [hunk 2]
```diff
@@ -1,14 +1,11 @@
...
-```
+For post training quantization, User can use `torchao` directly: https://github.com/pytorch/ao/blob/main/torchao/quantization/README.md to quantize their model
```
Comment (ebsmothers) [inline recipes/quantization.md:5]:
```suggestion
For post training quantization, we recommend using `torchao` directly: https://github.com/pytorch/ao/blob/main/torchao/quantization/README.md to quantize their model
```

**Review by ebsmothers — COMMENTED** [review]
Can you also update our live docs to account for this? See e.g. [here](https://pytorch.org/torchtune/main/tutorials/e2e_flow.html#speeding-up-generation-using-quantization) and [here](https://pytorch.org/torchtune/main/tutorials/llama3.html#faster-generation-via-quantization)

File: recipes/quantization.md (hunk 1 above)
Comment (jerryzh168) [reply recipes/quantization.md:8]: should we just link to QAT tutorial directly here?

**Comment by jerryzh168** [comment]
> Can you also update our live docs to account for this? See e.g. [here](https://pytorch.org/torchtune/main/tutorials/e2e_flow.html#speeding-up-generation-using-quantization) and [here](https://pytorch.org/torchtune/main/tutorials/llama3.html#faster-generation-via-quantization)

should I remove quantization walk through and just point to torchao there?

File: recipes/quantization.md (hunk 1 above)
Comment (jerryzh168) [reply recipes/quantization.md:8]: also I thought you are using `Int8DynActInt4WeightQuantizer` but not the quantization receipe?

**Review by ebsmothers — COMMENTED** [review]
Does this also mean that we no longer support generation with quantized models? E.g. what about [this portion](https://github.com/pytorch/torchtune/blob/651b18cb2b4154305c3f9f9a0f138dfd0da04699/recipes/generate.py#L40-L41) of our generation recipe? Is it possible to just load in the AO quantized checkpoint directly or no?

**Comment by jerryzh168** [comment]
> Does this also mean that we no longer support generation with quantized models? E.g. what about [this portion](https://github.com/pytorch/torchtune/blob/651b18cb2b4154305c3f9f9a0f138dfd0da04699/recipes/generate.py#L40-L41) of our generation recipe? Is it possible to just load in the AO quantized checkpoint directly or no?

yeah we'll remove this part as well, but currently it's kept for the last PTQ quantizer Int8DynActInt4WeightQuantizer, that will be removed after the QAT refactor. cc @andrewor14

**Comment by jerryzh168** [comment]
we expect people to run accuracy eval and perf benchmark in torchao as well (see updated doc as well)

**Comment by jerryzh168** [comment]
> Is it possible to just load in the AO quantized checkpoint directly or no?

I think this could be yes, but I haven't tested in torchtune, here are some tests in torchao: https://github.com/pytorch/ao/blob/e5df48e1033611d6433dc3fd1de7923e0f555ebc/test/quantization/test_quant_api.py#L643-L677

File: docs/source/tutorials/e2e_flow.rst [hunk 3]
```diff
@@ -319,126 +319,9 @@ Bay Area!
 Speeding up Generation using Quantization
 -----------------------------------------
 
-We saw that the generation recipe took around 11.6 seconds to generate 300 tokens.
-One technique commonly used to speed up inference is quantization. torchtune provides
-an integration with the `TorchAO <https://github.com/pytorch-labs/ao>`_
-quantization APIs. Let's first quantize the model using 4-bit weights-only quantization
-and see if this improves generation speed.
+We rely on `TorchAO <https://github.com/pytorch-labs/ao>`_ for `post training quantization <https://github.com/pytorch/ao/tree/main/torchao/quantization#quantization>`_, please take the fine tuned model from torchtune and quantize them directly in `TorchAO`, you can also run `eval/benchmark <https://github.com/pytorch/ao/tree/main/torchao/_models/llama>`_. in torchao as well.
```
Comment (joecummings) [inline docs/source/tutorials/e2e_flow.rst:322]: Can we include the few lines of code needed to do this with AO? Would make our tutorial way more clear.

File: docs/source/tutorials/llama3.rst [hunk 4]
```diff
@@ -238,104 +238,10 @@ Faster generation via quantization
 ----------------------------------
 
 We can see that the model took just under 11 seconds, generating almost 19 tokens per second.
-We can speed this up a bit by quantizing our model. Here we'll use 4-bit weights-only quantization
-as provided by `torchao <https://github.com/pytorch-labs/ao>`_.
+We rely on `TorchAO <https://github.com/pytorch-labs/ao>`_ for `post training quantization <https://github.com/pytorch/ao/tree/main/torchao/quantization#quantization>`_, please take the fine tuned model from torchtune and quantize them directly in `TorchAO`, you can also run `eval/benchmark <https://github.com/pytorch/ao/tree/main/torchao/_models/llama>`_. in torchao as well.
```
Comment (joecummings) [inline docs/source/tutorials/llama3.rst:241]: Same comment as above.

File: docs/source/tutorials/e2e_flow.rst (hunk 3 above)
Comment (jerryzh168) [reply docs/source/tutorials/e2e_flow.rst]: added

**Review by ebsmothers — CHANGES_REQUESTED** [review]
Need to fix the live docs. E.g.

Screenshot 2024-07-26 at 4 04 37 PM

**Review by ebsmothers — APPROVED** [review]
